test/topology/test_topology_failure_recovery.py

```diff
@@ -4,73 +4,100 @@
+
+        This helps to avoid making any assumptions on the order of the servers,
+        as in general the order of the cluster running servers can be somewhat
+        random (especially when initialized in parallel).
+    """
+    servers = await manager.running_servers()
+    random.shuffle(servers)
+    return servers
+
+
 async def inject_error_on(manager, error_name, servers):
-    errs = [manager.api.enable_injection(s.ip_addr, error_name, True) for s in servers]
+    """ Inject an error on the given servers. """
+    errs = [manager.api.enable_injection(s.ip_addr, error_name, one_shot=True) for s in servers]
     await asyncio.gather(*errs)
 
+
 @pytest.mark.asyncio
 @skip_mode('release', 'error injections are not supported in release mode')
 async def test_topology_streaming_failure(request, manager: ManagerClient):
     """Fail streaming while doing a topology operation"""
+    init_random_seed()
     # decommission failure
-    servers = await manager.running_servers()
+    servers = await get_running_servers(manager)
     logs = [await manager.server_open_log(srv.server_id) for srv in servers]
     marks = [await log.mark() for log in logs]
     await manager.api.enable_injection(servers[2].ip_addr, 'stream_ranges_fail', one_shot=True)
     await manager.decommission_node(servers[2].server_id, expected_error="Decommission failed. See earlier errors")
-    servers = await manager.running_servers()
+    servers = await get_running_servers(manager)
     assert len(servers) == 3
     matches = [await log.grep("raft_topology - rollback.*after decommissioning failure, moving transition state to rollback to normal",
                from_mark=mark) for log, mark in zip(logs, marks)]
     assert sum(len(x) for x in matches) == 1
     # bootstrap failure
     marks = [await log.mark() for log in logs]
-    servers = await manager.running_servers()
+    servers = await get_running_servers(manager)
     s = await manager.server_add(start=False, config={
         'error_injections_at_startup': ['stream_ranges_fail']
     })
     await manager.server_start(s.server_id, expected_error="Bootstrap failed. See earlier errors")
-    servers = await manager.running_servers()
+    servers = await get_running_servers(manager)
     assert s not in servers
     matches = [await log.grep("raft_topology - rollback.*after bootstrapping failure, moving transition state to left token ring",
                from_mark=mark) for log, mark in zip(logs, marks)]
     assert sum(len(x) for x in matches) == 1
     # bootstrap failure in raft barrier
     marks = [await log.mark() for log in logs]
-    servers = await manager.running_servers()
+    servers = await get_running_servers(manager)
     s = await manager.server_add(start=False)
-    await manager.api.enable_injection(servers[1].ip_addr, 'raft_topology_barrier_fail', one_shot=True)
+    await inject_error_on(manager, "raft_topology_barrier_fail", servers)
```

Comment:
Hm, this is tricky. I see a potential problem here.

The error is injected on all servers -- good. But then we know that one server will not stumble on it -- the coordinator. Which means that it will remain armed on the coordinator. Now suppose the coordinator changes. The previous coordinator has still the injection armed. This might potentially interfere with the remainder of this test -- e.g. one of these operations might stumble on this injection, not on the other ones, and either fail with unexpected error, or pass but not because it tested what it was supposed to test.

To eliminate this possibility and be 100% safe, I think you should explicitly disable this injection after this case finishes (i.e. before `#rebuild failure` below)

Reply:
Yes, that sounds right. Added the explicit removal of the error injection.